src/qat/purr/compiler/instructions.py

```diff
@@ -741,3 +752,155 @@ def calculate_duration(instruction, return_samples: bool = True):
         calc_sample = block_number * block_time
 
     return calc_sample
+
+
+class InstructionBlock:
+    """An Instruction grouping type. Allows working with blocks of Instructions as a
+    unit."""
+
+    instructions: List[Instruction]
+
+    def _validate_targets(self, targets, valid_types):
+        if targets is None:
+            targets = []
+        elif not isinstance(targets, List):
+            targets = [targets]
+
+        invalid_targets = [val for val in targets if not isinstance(val, valid_types)]
+        if any(invalid_targets):
+            invalid_targets_str = ",".join([str(val) for val in invalid_targets])
+            raise ValueError(f"Invalid targets for {type(self)}: {invalid_targets_str}")
+
+        return targets
+
+
+class QuantumInstructionBlock(InstructionBlock):
```

Comment:
mimics the instruction builder structure, could the instruction builder be reused in this case? 
after all it was designed to support instruction blocks as you seem to use them:

```
    def add(
        self,
        components: Union[
            "InstructionBuilder",
            Instruction,
            List[Union["InstructionBuilder", Instruction]],
        ],
    ):
        pass
```

Reply:
Short answer. No. The instruction builder can hold a list of instructions but is does not support the other features of `InstructionBlock`.